### ISSUE TITLE
A more elegant implement for temporary table

### ISSUE THREAD

**Issue by lcwangchao** [pingcap/tidb#26952]
## Enhancement

Temporary tables do not read/write TiKV. For global temporary table, it has an empty snapshot data and for local temporary table, it keeps snapshot data in session. Currently, snapshot does not support reading from memory, so when reading a temporary table's snapshot record, we need to do some tricks.

For example in point get, when reading snapshot, just get the record from session (nil record if it is a global temporary table).

https://github.com/pingcap/tidb/blob/300f15959d480c277e788311d2fab3e32af65bd6/executor/point_get.go#L405-L429

For batch_point_get, we make a custom snapshot `temporaryTableSnapshot` implements `kv.Snapshot`  to read temporary session data.

https://github.com/pingcap/tidb/blob/300f15959d480c277e788311d2fab3e32af65bd6/executor/batch_point_get.go#L149-L169

The above examples are not elegant, that is because we must add special codes in EVERY places where we need to access snapshot data. If we forget to do this, bug occurs.

A better practice is to make the `kv.Snapshot` more generic. That is, when the key is belongs to a temporary table, it read it from the session, otherwise from TiKV. After that, we can delete a lot of codes like `if tblInfo.TempTableType != model.TempTableNone` and to make a unified snapshot access for both temporary and normal table.

One difficulty to achieve it is that Snapshot object does not have the context of the table currently reading. So it can not infer whether to get the data from memory or not. In fact, snapshot does not have any concept for table, it is a more underlaying utility to access bytes key and values. We need to introduce some different abstractions to snapshot to make it.

Steps:

- [x] #27541
- [x] #27917
- [x] #27775
- [x] #27817
- [x] #28035
- [x] #28142
- [x] #28217

Below pull requests are outdated and not used by new design

- #27626
- #27587
- #27628

**Comment by lcwangchao** [comment]
I posted a draft here: https://github.com/tikv/client-go/pull/262 to make KVSnapshot able to read data from memory. It introduces two methods `SetMemSnapshotDB` to determine where it should read memory data from and `SetSortedMemScanRanges` to determine key ranges for memory reading.

I posted another draft in tidb: https://github.com/pingcap/tidb/pull/26948 to use above Snapshot to access temporary table snapshot data.
